### PR TITLE
Fix sentence endings.

### PR DESCRIPTION
**Pull request by sarunasb**
When something was NOT successfully done, shouldn't that message be emphasized?...

### CODE REVIEW

**Comment by Alex-Jordan** [comment]
This is more complicated than it looks. `maketext` accepts a string, and then uses the localization files in `lib/WeBWorK/Localize` to look up how to say that in whatever the local language is. By adding the period here, now the localization files would not have the string they need to look up.

You could edit all the localization files, or just the English one(s). Or make an edit both here in ProblemUtil.pm and in the localization files. But you should also check if anywhere else the same string happens to be passed to `maketext`.

**Comment by sarunasb** [comment]
The sentences are in Localize files and three places in the code, GatewayQuiz.pm (1) and ProblemUtil.pm (2).

`webwork2-dev$ grep -R 'Your score was successfully sent to the LMS' *
lib/WeBWorK/Localize/ko.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/en.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/en.po:msgstr "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/heb.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/es.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/fr.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/ru_RU.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/zh_hk.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/en_us.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/en_us.po:msgstr "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/tr.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/zh_CN.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/hu.po:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/Localize/webwork2.pot:msgid "Your score was successfully sent to the LMS"
lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:                        $r->maketext("Your score was successfully sent to the LMS") :
lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:                                      $scoreRecordedMessage .= $r->maketext("Your score was successfully sent to the LMS");
lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:                                      $scoreRecordedMessage .= $r->maketext("Your score was successfully sent to the LMS");

webwork2-dev$ grep -R 'Your score was not successfully sent to the LMS' *
lib/WeBWorK/Localize/ko.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/en.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/en.po:msgstr "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/heb.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/es.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/fr.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/ru_RU.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/zh_hk.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/en_us.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/en_us.po:msgstr "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/tr.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/zh_CN.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/hu.po:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/Localize/webwork2.pot:msgid "Your score was not successfully sent to the LMS"
lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:                        $r->maketext("Your score was not successfully sent to the LMS");
lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:                                      $scoreRecordedMessage .= $r->maketext("Your score was not successfully sent to the LMS");
lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:                                      $scoreRecordedMessage .= $r->maketext("Your score was not successfully sent to the LMS");`

Loking at webwork2.pot, msgid for complete sentences are all terminated with dots. 

Would it be correct to:
1) correct msgid entry in webwork2.pot;
2) all corresponding msgid and msgstr in language files;
3) text in maketext() for occurences in GatewayQuiz and ProblemUtil?

**Comment by Alex-Jordan** [comment]
Yes, that sounds good. If you add those edits to this pull request, I'll review it. I'd like one more reviewer, especially someone more familiar with localization. But yes, what you outline sounds right. 

If you are willing, would you look more closely through `webwork2.pot` and see if any other sentences need punctuation? It could all go together here.

**Comment by drgrice1** [comment]
I know there is one in WeBWorK/ContentGenerator/ShowMeAnother.pm on line 474.  That one has bugged me for a while.  I have a branch on my fork (https://github.com/drgrice1/webwork2/tree/fix-showmeanother-missing-period) that fixes this one.  You can look at that if you want to see the correct way to fix these strings.  Although, I think you have it covered in what you said earlier.

**Comment by taniwallach** [comment]
For the long term - we would also want to try to set up proper procedures to sync changes to the `msg_id` between the files and Transifex. So far all the work on coordinating changes between the the code, the POT/PO files and Transifex has been to the best of my knowledge very ad-hoc.

About 2 years ago, after a big Hebrew translation effort done in Transifex - I worked on getting all the data from Transifex into the PO files, as well as to update the main POT files. It took some effort to create updated files on the WW side (to get the new strings from the code into the translation files), merge in the data from the PO files from Transifex and WW, and the fix up some issues which came up. (I do not think I tried to update the POT file on Transifex, as I was very new to all of this stuff back then.) 

One of the bigger issues was that there are some special `msgid` values which are placeholders for which English needs a "translation". The problems which occurred due to this were fixed in https://github.com/openwebwork/webwork2/commit/7f7bb3889b33cba10b9515d0af6198cedd984071#diff-46bc9c1b52dfa8d7b02e4e0624d4df59 . If I recall/understand correctly, the issue is that Transifex does not expect/support "translations" to be different from the `msg_id` in the source languages. If that is the case - a better mechanism to avoid needing to keep putting these special messages back in to the English PO files might be to have the Perl code check if the special `msg_id` is returned, and in that case put in the English text.

See: https://github.com/openwebwork/webwork2/commits/master/lib/WeBWorK/Localize/en.po

**Comment by sarunasb** [comment]
Added dots to "...successfully sent to the LMS" in language files, GatewayQuiz and ProblemUtil.

**Comment by robert-marik** [comment]
FWIW, I managed few years ago a project in 15 languages. Once the source strings were updated, I prepared new pot file and merged with mkvmerge tool with existing translations. Minor editions were paired and marked as fuzzy and it was easy for translator to fix them. Especially after some simple edits like dot at the end.

**Comment by taniwallach** [comment]
> FWIW, I managed few years ago a project in 15 languages. Once the source strings were updated, I prepared new pot file and merged with mkvmerge tool with existing translations. Minor editions were paired and marked as fuzzy and it was easy for translator to fix them. Especially after some simple edits like dot at the end.

@robert-marik Maybe you help us develop a good working process for localization updates... Please also submit you Czech translation files as a pull request.

File: lib/WeBWorK/Localize/heb.po [hunk 1]
```diff
@@ -8440,7 +8440,7 @@ msgstr "הניקוד שלך לא נשמר."
 #: /opt/webwork/webwork2/lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:1892
 #: /opt/webwork/webwork2/lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:290
 #: /opt/webwork/webwork2/lib/WeBWorK/ContentGenerator/ProblemUtil/ProblemUtil.pm:296
-msgid "Your score was not successfully sent to the LMS"
+msgid "Your score was not successfully sent to the LMS."
 msgstr "הניקוד שלך לא נשלח בהצלחה ל-LMS"
```
Comment (Alex-Jordan) [inline lib/WeBWorK/Localize/heb.po:8444]: Adding a period to the English msgid here. @taniwallach, does the Hebrew need the same punctuation added?

Comment (Alex-Jordan) [reply lib/WeBWorK/Localize/heb.po]: A friend who knows Hebrew told me what to do here, and I added a commit to the remote branch.

**Comment by Alex-Jordan** [comment]
I would merge this but I'd feel a little better if @taniwallach thinks I did the right thing with the Hebrew localization. I'm confused by the right-to-left aspect of this. As I look at the string here in GitHub, the words are in backwards order according to my Hebrew literate friend. But I'm guessing that in the right environment, they would be presented in reverse order, and then it's right for the end user. It all makes me feel iffy that the period is in the right place.

**Comment by dlglin** [comment]
When I look at the Hebrew in GitHub the words are in the correct RTL order.  If the strings are presented as they appear here, then "-LMS" is at the wrong end of the sentence.  It should be "LMS-הניקוד שלך לא נשלח בהצלחה ל", unless the RTL code does something weird with transposing words.

Also, I don't know if there is a Hebrew phrase for LMS that @taniwallach can provide.

**Comment by Alex-Jordan** [comment]
Yes, this is hard to understand, in part because I think web browsers are doing their own thing when they perceive the rtl script characters. If you go to [one of these lines with mixed Hebrew and Latin](https://github.com/openwebwork/webwork2/blob/develop/lib/WeBWorK/Localize/heb.po#L8444), try to mouse-click-drag-highlight that line of text starting from the far left of the line where `msgstr` is. For me, using Firefox, as soon as I reach the Hebrew characters, it highlights all of them and then highlights fewer and fewer as I move right. This is happening even though there is no HTML markup separating the Hebrew characters. The web browser still knows to start with `msgstr` going ltr, then jump to the far right of the Hebrew and go rtl, then return to the `-LMS` going ltr.

So it's just hard for me to tell if you are not a regular user of rtl (and in particular, of mixed rtl and ltr) how this is supposed to be put into the localization file.

**Comment by Alex-Jordan** [comment]
I have confirmed that putting the period at the far right of the string, to the right of "LMS" does the right thing in output. I used Hebrew as the language and tried a variety of direction settings. They all came out like:
```
.LMS-[Hebrew part here]
```
With the Hebrew's leftmost word being ל as expected. (I'm unsure about pasting the whole thing here because I wonder if a local web browser messes with its display order.)

**Pull request merged**